**Breast Cancer Detection**

**Team Members:**

Sajjani P - 1RVU22BSC087

Akanksha K - 1RVU22BSC039

Chaitanya N - 1RVU22BSC017

Sumedha VK - 1RVU22BSC101

**`Data Preprocessing`**




In [ ]:
import pandas as pd
df=pd.read_csv("Brest_cancer.csv")
df

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0    ...          17.33           184.60      2019.0           0.16220   
1    ...          23.41           158.80      1956.0           0.12380   
2    ...          25.53           152.50      1709.0           0.14440   
3    ...          26.50            98.87       567.7           0.20980   
4    ...          16.67           152.20      1575.0           0.13740   
..   ...            ...              ...         ...               ...   
564  ...          26.40           166.10      2027.0           0.14100   
565  ...          38.25           155.00      1731.0           0.11660   
566  ...          34.12           126.70      1124.0           0.11390   
567  ...          39.42           184.60      1821.0           0.16500   
568  ...          30.37            59.16       268.6           0.08996   

     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0              0.66560           0.7119                0.2654          0.4601   
1              0.18660           0.2416                0.1860          0.2750   
2              0.42450           0.4504                0.2430          0.3613   
3              0.86630           0.6869                0.2575          0.6638   
4              0.20500           0.4000                0.1625          0.2364   
..                 ...              ...                   ...             ...   
564            0.21130           0.4107                0.2216          0.2060   
565            0.19220           0.3215                0.1628          0.2572   
566            0.30940           0.3403                0.1418          0.2218   
567            0.86810           0.9387                0.2650          0.4087   
568            0.06444           0.0000                0.0000          0.2871   

     fractal_dimension_worst  Unnamed: 32  
0                    0.11890          NaN  
1                    0.08902          NaN  
2                    0.08758          NaN  
3                    0.17300          NaN  
4      

**Data Cleaning**

In [ ]:
df.isnull().sum()

id                           0
diagnosis                    0
radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed:

In [ ]:
df.isnull().sum()/len(df)*100

id                           0.0
diagnosis                    0.0
radius_mean                  0.0
texture_mean                 0.0
perimeter_mean               0.0
area_mean                    0.0
smoothness_mean              0.0
compactness_mean             0.0
concavity_mean               0.0
concave points_mean          0.0
symmetry_mean                0.0
fractal_dimension_mean       0.0
radius_se                    0.0
texture_se                   0.0
perimeter_se                 0.0
area_se                      0.0
smoothness_se                0.0
compactness_se               0.0
concavity_se                 0.0
concave points_se            0.0
symmetry_se                  0.0
fractal_dimension_se         0.0
radius_worst                 0.0
texture_worst                0.0
perimeter_worst              0.0
area_worst                   0.0
smoothness_worst             0.0
compactness_worst            0.0
concavity_worst              0.0
concave points_worst         0.0
symmetry_w

In [ ]:
df.drop(['Unnamed: 32'],axis = 1 ,inplace = True) # Dropping columns with null values
df

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0    ...        25.380          17.33           184.60      2019.0   
1    ...        24.990          23.41           158.80      1956.0   
2    ...        23.570          25.53           152.50      1709.0   
3    ...        14.910          26.50            98.87       567.7   
4    ...        22.540          16.67           152.20      1575.0   
..   ...           ...            ...              ...         ...   
564  ...        25.450          26.40           166.10      2027.0   
565  ...        23.690          38.25           155.00      1731.0   
566  ...        18.980          34.12           126.70      1124.0   
567  ...        25.740          39.42           184.60      1821.0   
568  ...         9.456          30.37            59.16       268.6   

     smoothness_worst  compactness_worst  concavity_worst  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                  0.2654          0.4601                  0.11890  
1                  0.1860          0.2750                  0.08902  
2                  0.2430          0.3613                  0.08758  
3                  0.2575          0.6638                  0.17300  
4                  0.1625          0.2364                  0.07678  
..                    ...             ...                      ...  
564                0.2216       

In [ ]:
df.duplicated().sum()

0

In [ ]:
df.describe()

id  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  5.690000e+02   569.000000    569.000000      569.000000   569.000000   
mean   3.037183e+07    14.127292     19.289649       91.969033   654.889104   
std    1.250206e+08     3.524049      4.301036       24.298981   351.914129   
min    8.670000e+03     6.981000      9.710000       43.790000   143.500000   
25%    8.692180e+05    11.700000     16.170000       75.170000   420.300000   
50%    9.060240e+05    13.370000     18.840000       86.240000   551.100000   
75%    8.813129e+06    15.780000     21.800000      104.100000   782.700000   
max    9.113205e+08    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count     569.000000  ...    569.000000     569.000000       569.000000   
mean        0.181162  ...     16.269190      25.677223       107.261213   
std         0.027414  ...      4.833242       6.146258        33.602542   
min         0.106000  ...      7.930000      12.020000        50.410000   
25%         0.161900  ...     13.010000      21.080000        84.110000   
50%         0.179200  ...     14.970000      25.410000        97.660000   
75%         0.195700  ...     18.790000      29.720000       125.400000   
max         0.304000  ...     36.040000      49.540000       251.200000   

        area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count   569.000000        569.000000         569.000000       569.000000   
mean    880.583128          0.132369           0.254265         0.272188   
std     569.356993          0.022832           0.157336         0.208624   
min     185.200000          0.071170           0.027290         0.000000   
25%     515.300000          0.116600           0.147200         0.114500   
50%     686.500000          0.131300           0.211900         0.226700   
75%    1084.000000          0.146000           0.339100         0.382900   
max    4254.000000          0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 31 columns]

In [ ]:
df.isnull().sum()/len(df)*100

id                         0.0
diagnosis                  0.0
radius_mean                0.0
texture_mean               0.0
perimeter_mean             0.0
area_mean                  0.0
smoothness_mean            0.0
compactness_mean           0.0
concavity_mean             0.0
concave points_mean        0.0
symmetry_mean              0.0
fractal_dimension_mean     0.0
radius_se                  0.0
texture_se                 0.0
perimeter_se               0.0
area_se                    0.0
smoothness_se              0.0
compactness_se             0.0
concavity_se               0.0
concave points_se          0.0
symmetry_se                0.0
fractal_dimension_se       0.0
radius_worst               0.0
texture_worst              0.0
perimeter_worst            0.0
area_worst                 0.0
smoothness_worst           0.0
compactness_worst          0.0
concavity_worst            0.0
concave points_worst       0.0
symmetry_worst             0.0
fractal_dimension_worst    0.0
dtype: f

In [ ]:
df.duplicated().sum()

0

In [ ]:
df.to_csv('Brest_cancer_cleaned.csv',index=False)#saving data frame into csv

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
df = pd.read_csv('Brest_cancer_cleaned.csv')

In [ ]:
df.isnull().sum()/len(df)*100

id                         0.0
radius_mean                0.0
texture_mean               0.0
perimeter_mean             0.0
area_mean                  0.0
smoothness_mean            0.0
compactness_mean           0.0
concavity_mean             0.0
concave points_mean        0.0
symmetry_mean              0.0
fractal_dimension_mean     0.0
radius_se                  0.0
texture_se                 0.0
perimeter_se               0.0
area_se                    0.0
smoothness_se              0.0
compactness_se             0.0
concavity_se               0.0
concave points_se          0.0
symmetry_se                0.0
fractal_dimension_se       0.0
radius_worst               0.0
texture_worst              0.0
perimeter_worst            0.0
area_worst                 0.0
smoothness_worst           0.0
compactness_worst          0.0
concavity_worst            0.0
concave points_worst       0.0
symmetry_worst             0.0
fractal_dimension_worst    0.0
diagnosis                  0.0
dtype: f

**Encoding**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   radius_mean              569 non-null    float64
 2   texture_mean             569 non-null    float64
 3   perimeter_mean           569 non-null    float64
 4   area_mean                569 non-null    float64
 5   smoothness_mean          569 non-null    float64
 6   compactness_mean         569 non-null    float64
 7   concavity_mean           569 non-null    float64
 8   concave points_mean      569 non-null    float64
 9   symmetry_mean            569 non-null    float64
 10  fractal_dimension_mean   569 non-null    float64
 11  radius_se                569 non-null    float64
 12  texture_se               569 non-null    float64
 13  perimeter_se             569 non-null    float64
 14  area_se                  5

Label encoding


In [ ]:
# Using label encoding

from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

# Fit and transform the data
df['diagnosis'] = le.fit_transform(df['diagnosis'])

In [ ]:
df

id  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302        17.99         10.38          122.80     1001.0   
1      842517        20.57         17.77          132.90     1326.0   
2    84300903        19.69         21.25          130.00     1203.0   
3    84348301        11.42         20.38           77.58      386.1   
4    84358402        20.29         14.34          135.10     1297.0   
..        ...          ...           ...             ...        ...   
564    926424        21.56         22.39          142.00     1479.0   
565    926682        20.13         28.25          131.20     1261.0   
566    926954        16.60         28.08          108.30      858.1   
567    927241        20.60         29.33          140.10     1265.0   
568     92751         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     symmetry_mean  ...  texture_worst  perimeter_worst  area_worst  \
0           0.2419  ...          17.33           184.60      2019.0   
1           0.1812  ...          23.41           158.80      1956.0   
2           0.2069  ...          25.53           152.50      1709.0   
3           0.2597  ...          26.50            98.87       567.7   
4           0.1809  ...          16.67           152.20      1575.0   
..             ...  ...            ...              ...         ...   
564         0.1726  ...          26.40           166.10      2027.0   
565         0.1752  ...          38.25           155.00      1731.0   
566         0.1590  ...          34.12           126.70      1124.0   
567         0.2397  ...          39.42           184.60      1821.0   
568         0.1587  ...          30.37            59.16       268.6   

     smoothness_worst  compactness_worst  concavity_worst  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  diagnosis  
0                  0.2654          0.4601                  0.11890          1  
1                  0.1860          0.2750                  0.08902          1  
2                  0.2430          0.3613                  0.08758          1  
3                  0.2575          0.6638                  0.17300          1  
4                  0.1625          0.2364                  0.07678          1  
..                    ...             ...                      ...        ...  
564                0.2216          0.2060                  0.07

In [ ]:
# Labels are the index values of the elements present in the below array

le.classes_

array(['B', 'M'], dtype=object)

In [ ]:
X = df.iloc[::,:-1]
y = df.iloc[::,-1:]

In [ ]:
print (y)

     diagnosis
0            1
1            1
2            1
3            1
4            1
..         ...
564          1
565          1
566          1
567          1
568          0

[569 rows x 1 columns]


Scaling


In [ ]:
from sklearn.preprocessing import MinMaxScaler

minmax = MinMaxScaler()
minmax_data = minmax.fit_transform(X)
minmax_df = pd.DataFrame(minmax_data, columns=X.columns)
minmax_df

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_data.py:473: RuntimeWarning: All-NaN slice encountered
  data_min = np.nanmin(X, axis=0)
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_data.py:474: RuntimeWarning: All-NaN slice encountered
  data_max = np.nanmax(X, axis=0)


id  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    0.000915     0.521037      0.022658        0.545989   0.363733   
1    0.000915     0.643144      0.272574        0.615783   0.501591   
2    0.092495     0.601496      0.390260        0.595743   0.449417   
3    0.092547     0.210090      0.360839        0.233501   0.102906   
4    0.092559     0.629893      0.156578        0.630986   0.489290   
..        ...          ...           ...             ...        ...   
564  0.001007     0.690000      0.428813        0.678668   0.566490   
565  0.001007     0.622320      0.626987        0.604036   0.474019   
566  0.001008     0.455251      0.621238        0.445788   0.303118   
567  0.001008     0.644564      0.663510        0.665538   0.475716   
568  0.000092     0.036869      0.501522        0.028540   0.015907   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0           0.593753          0.792037        0.703140             0.731113   
1           0.289880          0.181768        0.203608             0.348757   
2           0.514309          0.431017        0.462512             0.635686   
3           0.811321          0.811361        0.565604             0.522863   
4           0.430351          0.347893        0.463918             0.518390   
..               ...               ...             ...                  ...   
564         0.526948          0.296055        0.571462             0.690358   
565         0.407782          0.257714        0.337395             0.486630   
566         0.288165          0.254340        0.216753             0.263519   
567         0.588336          0.790197        0.823336             0.755467   
568         0.000000          0.074351        0.000000             0.000000   

     symmetry_mean  ...  perimeter_worst  area_worst  smoothness_worst  \
0         0.686364  ...         0.668310    0.450698          0.601136   
1         0.379798  ...         0.539818    0.435214          0.347553   
2         0.509596  ...         0.508442    0.374508          0.483590   
3         0.776263  ...         0.241347    0.094008          0.915472   
4         0.378283  ...         0.506948    0.341575          0.437364   
..             ...  ...              ...         ...               ...   
564       0.336364  ...         0.576174    0.452664          0.461137   
565       0.349495  ...         0.520892    0.379915          0.300007   
566       0.267677  ...         0.379949    0.230731          0.282177   
567       0.675253  ...         0.668310    0.402035          0.619626   
568       0.266162  ...         0.043578    0.020497          0.124084   

     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.619292         0.568610              0.912027        0.598462   
1             0.154563         0.192971              0.639175        0.233590   
2             0.385375         0.359744              0.835052        0.403706   
3             0.814012         0.548642              0.884880        1.000000   
4             0.172415         0.319489              0.558419        0.157500   
..                 ...              ...                   ...             ...   
564           0.178527         0.328035              0.761512        0.097575   
565           0.159997         0.256789              0.559450        0.198502   
566           0.273705         0.271805              0.487285        0.128721   
567           0.815758         0.749760              0.910653        0.497142   
568           0.036043         0.000000              0.000000        0.257441   

     fractal_dimension_worst  Unnamed: 31  Unnamed: 32  
0                   0.418864          NaN          NaN  
1                   0.222878          NaN          NaN  
2                   0.213433          NaN          NaN  
3                   0.773711          NaN          NaN  
4                   0.142595          NaN          NaN  
..   

In [ ]:
minmax_df.drop(['Unnamed: 31','Unnamed: 32'],axis = 1 ,inplace = True)

In [ ]:
minmax_df

id  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    0.000915     0.521037      0.022658        0.545989   0.363733   
1    0.000915     0.643144      0.272574        0.615783   0.501591   
2    0.092495     0.601496      0.390260        0.595743   0.449417   
3    0.092547     0.210090      0.360839        0.233501   0.102906   
4    0.092559     0.629893      0.156578        0.630986   0.489290   
..        ...          ...           ...             ...        ...   
564  0.001007     0.690000      0.428813        0.678668   0.566490   
565  0.001007     0.622320      0.626987        0.604036   0.474019   
566  0.001008     0.455251      0.621238        0.445788   0.303118   
567  0.001008     0.644564      0.663510        0.665538   0.475716   
568  0.000092     0.036869      0.501522        0.028540   0.015907   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0           0.593753          0.792037        0.703140             0.731113   
1           0.289880          0.181768        0.203608             0.348757   
2           0.514309          0.431017        0.462512             0.635686   
3           0.811321          0.811361        0.565604             0.522863   
4           0.430351          0.347893        0.463918             0.518390   
..               ...               ...             ...                  ...   
564         0.526948          0.296055        0.571462             0.690358   
565         0.407782          0.257714        0.337395             0.486630   
566         0.288165          0.254340        0.216753             0.263519   
567         0.588336          0.790197        0.823336             0.755467   
568         0.000000          0.074351        0.000000             0.000000   

     symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.686364  ...      0.620776       0.141525         0.668310   
1         0.379798  ...      0.606901       0.303571         0.539818   
2         0.509596  ...      0.556386       0.360075         0.508442   
3         0.776263  ...      0.248310       0.385928         0.241347   
4         0.378283  ...      0.519744       0.123934         0.506948   
..             ...  ...           ...            ...              ...   
564       0.336364  ...      0.623266       0.383262         0.576174   
565       0.349495  ...      0.560655       0.699094         0.520892   
566       0.267677  ...      0.393099       0.589019         0.379949   
567       0.675253  ...      0.633582       0.730277         0.668310   
568       0.266162  ...      0.054287       0.489072         0.043578   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      0.450698          0.601136           0.619292         0.568610   
1      0.435214          0.347553           0.154563         0.192971   
2      0.374508          0.483590           0.385375         0.359744   
3      0.094008          0.915472           0.814012         0.548642   
4      0.341575          0.437364           0.172415         0.319489   
..          ...               ...                ...              ...   
564    0.452664          0.461137           0.178527         0.328035   
565    0.379915          0.300007           0.159997         0.256789   
566    0.230731          0.282177           0.273705         0.271805   
567    0.402035          0.619626           0.815758         0.749760   
568    0.020497          0.124084           0.036043         0.000000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.912027        0.598462                 0.418864  
1                0.639175        0.233590                 0.222878  
2                0.835052        0.403706                 0.213433  
3                0.884880        1.000000                 0.773711  
4                0.558419        0.157500                 0.142595  
..                    ...             ...

In [ ]:
X = minmax_df  # Features
y = df['diagnosis']

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.40)

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors = 15)
classifier.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=15)

In [ ]:
y_pred = classifier.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)
cr = classification_report(y_test, y_pred)
print("\nClassification Report:",)
print (cr)
acc = accuracy_score(y_test,y_pred)
print("\nAccuracy:",acc)

Confusion Matrix:
[[139   1]
 [  7  81]]

Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.99      0.97       140
           1       0.99      0.92      0.95        88

    accuracy                           0.96       228
   macro avg       0.97      0.96      0.96       228
weighted avg       0.97      0.96      0.96       228


Accuracy: 0.9649122807017544


##**Report**

                        




The Breast Cancer Wisconsin dataset is made up of features extracted from a digitalized image of a fine needle aspirate (FNA) of the breast tumor.There are 569 entries in the dataset. Independent variables include measures such as radius, texture, perimeter, and smoothness and the dependent variable includes tumors being classified as malignant or benign.

**Steps for ML project:**

**Data Preprocessing:**
*  Step 1: Data preprocessing and found out that our dataset has no null values and duplicate values.
*  Step 2: There is "Diagnosis" category column in the dataset. Since the output only contains two sorts of classes benign and malignant ,we have to execute label encoding in order to transform it into a numerical column. Similar to how label encoding gives each category a distinct integer number. Even though we should have used one-hot encoding for our nominal categorical column, we tried both and discovered that label-encoding was more efficient. Here Malignant (M) is considered as 1 and Benign (B) is considered as 0
* Step 3: In this project, applying min-max scaling to numeric features ensures that all variables are uniformly scaled between 0 and 1, which makes it easier for the model to understand and make predictions.

**Model Selection:**
This is a classification algorithm and The model is using KNN algorithm as we have learnt only KNN and we have more practice in using KNN

**Confusion Matrix:**

The confusion matrix provides a tabular summary of the model's predictions versus the actual labels in the test dataset.
The first row and first column represents the actual instances and predicted instances of the benign class (0).
The second row and second represents the actual instances and predicted instances of the malignant class (1).

**This confusion matrix indicates:**
True Negatives (TN): 139 instances were correctly predicted as class 0.
False Positives (FP): 1 instance was incorrectly predicted as class 1.
False Negatives (FN): 7 instances were incorrectly predicted as class 0.
True Positives (TP): 81 instances were correctly predicted as class 1.

The false prediction in the model might be because the dataset is hard to understand, or maybe because of the selection of algorithm.

**Classification Report:**

The classification report provides a summary of Evaluation metrics, including precision, recall, F1-score, and support, for each class (0 and 1).

**For each class, the report provides:**
Precision for benign (class 0) is 95% and for malignant (class 1) is 99%.
Recall for benign is 99% and for malignant is 92%.
F1-score for benign is 97% and for malignant is 95%.
Accuracy is 96.49%, indicating overall model performance.

The precision, recall, and F1-score for both benign (class 0) and malignant (class 1) cases are relatively high, indicating that the model is making accurate predictions for both classes.
The accuracy of 96.49% is also high, suggesting that the model's overall performance is good.


**Conclusion:**

The breast cancer prediction model is performing good. It's highly accurate and can effectively distinguish between benign and malignant cases with great precision. This indicates its reliability and ability to assist doctors in identifying cancerous tumors more accurately.








